File: visualstudio-extension/src/GitReview.Host/ReviewStateManager.cs
using GitReview.Domain;

namespace GitReview.Host;

/// <summary>
/// Refresh pipeline: version probe → status --porcelain → list/config when needed.
/// Sole source of ReviewState for the panel.
/// </summary>
public sealed class ReviewStateManager
{
    private readonly CliInvoker _cli;
    private readonly Func<IReadOnlyList<string>> _repoRoots;
    private readonly Func<string?> _gitReviewPath;
    private ReviewState _state = new(Situation.CliMissing);
    private readonly object _gate = new();
    private int _refreshGen;

    public event Action<ReviewState>? StateChanged;

    public ReviewState Current
    {
        get { lock (_gate) return _state; }
    }

    /// <summary>
    /// False until a refresh has resolved a state. The seed above is a placeholder,
    /// not an answer, and drawing it tells the reviewer the CLI is missing before
    /// anyone has looked -- so the panel waits instead. Same rule as the JetBrains
    /// service's hasResolvedState and the extension's empty webview.
    /// </summary>
    public bool HasResolved { get; private set; }

    public ReviewStateManager(
        CliInvoker cli,
        Func<IReadOnlyList<string>> repoRoots,
        Func<string?>? gitReviewPath = null)
    {
        _cli = cli;
        _repoRoots = repoRoots;
        _gitReviewPath = gitReviewPath ?? (() => null);
    }

    public async Task<ReviewState> RefreshAsync(CancellationToken ct = default)
    {
        var gen = Interlocked.Increment(ref _refreshGen);
        var roots = _repoRoots();
        var sole = SoleTarget.PickSoleTarget(roots);

        if (roots.Count > 1)
        {
            return Publish(new ReviewState(
                Situation.Error,
                Stderr: "Open a single-folder workspace that is a git repository. git review uses one root (like the CLI cwd); multi-root is not supported."), gen);
        }

        if (sole is null)
        {
            return Publish(new ReviewState(
                Situation.Error,
                Stderr: UserCopy.NoSoleRoot), gen);
        }

        // Version probe: `git review --version` (verb is the flag after "review").
        // Declaring the CLI missing takes EVIDENCE of absence, not any failure at all:
        // the first invocation of a startup is the one most likely to go wrong for
        // reasons that are not the CLI, and that is where the install screen landed on
        // top of an installed CLI. With no evidence the probe is retried and the panel
        // keeps waiting -- nothing has been published yet -- instead of answering.
        var ver = await _cli.InvokeAsync("--version", Array.Empty<string>(), sole, cancellationToken: ct)
            .ConfigureAwait(false);
        var verdict = CliProbe.VersionVerdict(ver.Stderr, ver.ExitCode, ver.ErrorCode, ver.TimedOut);
        for (var attempt = 0; verdict == CliVerdict.Unknown && attempt < CliProbe.CliProbeRetries; attempt++)
        {
            await Task.Delay(CliProbe.CliProbeRetryDelayMs, ct).ConfigureAwait(false);
            ver = await _cli.InvokeAsync("--version", Array.Empty<string>(), sole, cancellationToken: ct)
                .ConfigureAwait(false);
            verdict = CliProbe.VersionVerdict(ver.Stderr, ver.ExitCode, ver.ErrorCode, ver.TimedOut);
        }

        if (verdict != CliVerdict.Ok)
        {
            // A CLI that takes too long is the opposite of a CLI that is not there, and
            // offering "install it with npm" there sends the reviewer to install what is
            // already installed.
            return Publish(ver.TimedOut
                ? new ReviewState(Situation.Error, Stderr: VersionTimeout)
                : new ReviewState(
                    Situation.CliMissing,
                    Stderr: string.IsNullOrWhiteSpace(ver.Stderr) ? "not found" : ver.Stderr), gen);
        }

        var versionLine = CliMessage.FirstCliLine(ver.Stdout);
        if (string.IsNullOrEmpty(versionLine))
        {
            // Some builds print version on stderr or fail only on status; continue.
        }
        else if (CliVersion.IsOutdated(versionLine))
        {
            return Publish(new ReviewState(Situation.CliOutdated, Stderr: versionLine), gen);
        }

        var status = await _cli.InvokeAsync(
                "status", new[] { "--porcelain" }, sole, cancellationToken: ct)
            .ConfigureAwait(false);

        if (IsMissing(status))
        {
            return Publish(new ReviewState(
                Situation.CliMissing,
                Stderr: CliMessage.CliErrorText(status.Stderr, status.Stdout, "CLI not found")), gen);
        }

        if (status.TimedOut)
        {
            return Publish(new ReviewState(
                Situation.Error,
                Stderr: "status --porcelain timed out"), gen);
        }

        var exit = status.ExitCode;
        ReviewState next;

        if (exit == 3)
        {
            // Out of range: the walk cursor no longer meets the range, usually because
            // the reviewer committed on top of the staged diff. The CLI writes that to
            // stderr and prints no porcelain at all, so this must not be parsed -- it
            // used to be, which made the entry that the reviewer can act on ("undo the
            // commits with git reset --soft, or abort") come out of the panel as
            // "porcelain output has no state record" under a generic error.
            next = new ReviewState(Situation.OutOfRange, Stderr: status.Stderr);
        }
        else if (exit == 0)
        {
            try
            {
                var parsed = Porcelain.ParsePorcelain(status.Stdout);
                var hasConflict = parsed.Finish is not null;
                var situation = SituationIds.For(exit, hasConflict, hasFinishPending: false);
                next = new ReviewState(
                    Situation: situation,
                    State: parsed.State,
                    Entries: parsed.Entries,
                    Files: parsed.FilesList,
                    Subjects: parsed.Subjects,
                    Authors: parsed.Authors,
                    Base: parsed.Base,
                    Finish: parsed.Finish,
                    Readonly: parsed.Readonly,
                    KeysOnly: parsed.KeysOnly,
                    Draft: parsed.Draft,
                    DraftPath: parsed.DraftPath);
            }
            catch (Exception e)
            {
                // What the CLI said beats what the parser says: porcelain it could not
                // read is more often a CLI that already explained itself than a bug in
                // the tokenizer, and the reviewer can act on the former.
                next = new ReviewState(
                    Situation.Error,
                    Stderr: status.Stderr.Trim().Length > 0 ? status.Stderr : e.Message);
            }
        }
        else if (exit == 2)
        {
            // no-review: list + config
            var listTask = _cli.InvokeAsync("list", new[] { "--porcelain" }, sole, cancellationToken: ct);
            var configTask = _cli.InvokeAsync("config", new[] { "--porcelain" }, sole, cancellationToken: ct);
            await Task.WhenAll(listTask, configTask).ConfigureAwait(false);
            var list = await listTask.ConfigureAwait(false);
            var config = await configTask.ConfigureAwait(false);

            var branches = list.ExitCode == 0
                ? Porcelain.ParseListPorcelain(list.Stdout)
                : Array.Empty<BranchRecord>();
            // One invocation, two readings: the fixes branches travel in the same
            // output as the inventory, so the footer section costs no extra
            // process per refresh.
            var fixes = list.ExitCode == 0
                ? Porcelain.ParseListFixes(list.Stdout)
                : Array.Empty<FixesRecord>();
            EffectiveConfig? eff = null;
            IReadOnlyList<CandidateBranch>? candidates = null;
            IReadOnlyList<CandidateRemote>? remotes = null;
            IReadOnlyList<DraftRecord>? drafts = null;
            // The guides and the walkthrough row come from THIS report, not from
            // status: outside a review status is what exits 2, so the only verb
            // that can carry them is this one. Without them the empty state drew
            // a Walkthrough section with two buttons and nothing else, while the
            // other two clients drew the rows.
            IReadOnlyList<GuideRecord>? guides = null;
            WalkthroughRecord? walkthrough = null;
            if (config.ExitCode == 0)
            {
                try
                {
                    var cp = ConfigPorcelain.ParseConfigPorcelain(config.Stdout);
                    eff = cp.Config;
                    candidates = cp.Candidates;
                    remotes = cp.Remotes;
                    drafts = cp.Drafts;
                    guides = cp.Guides;
                    walkthrough = cp.Walkthrough;
                }
                catch { /* leave null */ }
            }

            var hasPending = branches.Any(b => b.Finish?.State == "pending");
            var situation = SituationIds.For(2, hasFinishConflict: false, hasFinishPending: hasPending);
            next = new ReviewState(
                Situation: situation,
                Branches: branches,
                Fixes: fixes,
                Config: eff,
                Candidates: candidates,
                Remotes: remotes,
                Drafts: drafts,
                Guides: guides,
                Walkthrough: walkthrough);
        }
        else
        {
            next = new ReviewState(
                Situation.Error,
                Stderr: CliMessage.CliErrorText(
                    status.Stderr, status.Stdout, "Something went wrong reading the review state."));
        }

        return Publish(next, gen);
    }

    /// <summary>
    /// A version probe that ran out of time: the CLI is there and slow, not absent, so
    /// what the panel says is where to look.
    /// </summary>
    private const string VersionTimeout =
        "`git review --version` did not finish in time and was stopped. "
        + "Run it in a terminal to see how long it takes. "
        + "See the git review pane of the Output window.";

    private static bool IsMissing(InvokeResult r) =>
        r.ErrorCode is not null
        || (r.ExitCode is not null and not 0
            && (r.Stderr.Contains("is not a git command", StringComparison.OrdinalIgnoreCase)
                || r.Stderr.Contains("not found", StringComparison.OrdinalIgnoreCase)
                || r.Stderr.Contains("ENOENT", StringComparison.OrdinalIgnoreCase)
                || r.ErrorCode is "Win32Exception" or "FileNotFoundException"));

    private ReviewState Publish(ReviewState state, int gen)
    {
        if (gen != Volatile.Read(ref _refreshGen)) return Current;
        HasResolved = true;
        lock (_gate) _state = state;
        StateChanged?.Invoke(state);
        return state;
    }
}

File: jetbrains-plugin/src/main/kotlin/com/ezevillo/gitreview/host/ReviewStateManager.kt
package com.ezevillo.gitreview.host

import com.ezevillo.gitreview.domain.CLI_PROBE_RETRIES
import com.ezevillo.gitreview.domain.CLI_PROBE_RETRY_DELAY_MS
import com.ezevillo.gitreview.domain.CliVerdict
import com.ezevillo.gitreview.domain.ReviewState
import com.ezevillo.gitreview.domain.Situation
import com.ezevillo.gitreview.domain.isOutdated
import com.ezevillo.gitreview.domain.parseConfigPorcelain
import com.ezevillo.gitreview.domain.parseListFixes
import com.ezevillo.gitreview.domain.parseListPorcelain
import com.ezevillo.gitreview.domain.parsePorcelain
import com.ezevillo.gitreview.domain.situationFor
import com.ezevillo.gitreview.domain.versionVerdict
import java.util.concurrent.atomic.AtomicLong

/**
 * Coalesced refresh of review state from CLI porcelain. No UI dependency.
 */
class ReviewStateManager(
    private val invoker: CliRunner,
    private val gitReviewPath: () -> String?,
) {
    private val generation = AtomicLong(0)
    @Volatile
    var current: ReviewState = ReviewState(situation = Situation.ERROR, stderr = "not refreshed yet")
        private set

    private var versionCheckedGeneration: Long = -1

    fun invalidateVersion() {
        versionCheckedGeneration = -1
    }

    /**
     * Full refresh. When [cwd] is null → multi-root / no-repo error situation.
     */
    fun refresh(cwd: String?): ReviewState {
        val gen = generation.incrementAndGet()
        if (cwd == null) {
            val state = ReviewState(
                situation = Situation.ERROR,
                stderr = MULTI_ROOT_ERROR,
            )
            if (gen == generation.get()) current = state
            return state
        }

        if (versionCheckedGeneration != gen || current.situation == Situation.CLI_MISSING) {
            // Declaring the CLI missing takes EVIDENCE of absence, not any
            // failure at all: the first invocation of a startup is the one most
            // likely to go wrong for reasons that are not the CLI, and that is
            // where the install screen landed on top of an installed CLI. With
            // no evidence the probe is retried and the panel keeps waiting --
            // it has published nothing yet -- instead of answering.
            var ver = invoker.invoke("--version", emptyList(), cwd)
            var verdict = versionVerdict(ver.stderr, ver.exitCode, ver.errorCode, ver.timedOut)
            var attempt = 0
            while (verdict == CliVerdict.UNKNOWN && attempt < CLI_PROBE_RETRIES) {
                attempt++
                Thread.sleep(CLI_PROBE_RETRY_DELAY_MS)
                ver = invoker.invoke("--version", emptyList(), cwd)
                verdict = versionVerdict(ver.stderr, ver.exitCode, ver.errorCode, ver.timedOut)
            }
            if (verdict != CliVerdict.OK) {
                // A CLI that takes too long is the opposite of a CLI that is not
                // there, and offering "install it with npm" there sends the
                // reviewer to install what is already installed.
                val state = if (ver.timedOut) {
                    ReviewState(
                        situation = Situation.ERROR,
                        stderr = VERSION_TIMEOUT,
                    )
                } else {
                    ReviewState(
                        situation = Situation.CLI_MISSING,
                        stderr = ver.stderr.ifBlank { "CLI not found" },
                    )
                }
                if (gen == generation.get()) {
                    current = state
                    versionCheckedGeneration = gen
                }
                return state
            }
            // No version line means nothing to compare: the CLI answered, and
            // the status that follows judges better than isOutdated(""), which
            // returns true.
            val version = ver.stdout.trim()
            if (version.isNotEmpty() && isOutdated(version)) {
                val state = ReviewState(
                    situation = Situation.CLI_OUTDATED,
                    stderr = "installed version: $version",
                )
                if (gen == generation.get()) {
                    current = state
                    versionCheckedGeneration = gen
                }
                return state
            }
            versionCheckedGeneration = gen
        }

        val status = invoker.invoke("status", listOf("--porcelain"), cwd)
        if (status.timedOut) {
            val state = ReviewState(
                situation = Situation.ERROR,
                stderr = "status --porcelain timed out",
            )
            if (gen == generation.get()) current = state
            return state
        }

        var branches = emptyList<com.ezevillo.gitreview.domain.BranchRecord>()
        var fixes = emptyList<com.ezevillo.gitreview.domain.FixesRecord>()
        var config: com.ezevillo.gitreview.domain.EffectiveConfig? = null
        var candidates: List<com.ezevillo.gitreview.domain.CandidateBranch>? = null
        var remotes: List<com.ezevillo.gitreview.domain.CandidateRemote>? = null
        var drafts: List<com.ezevillo.gitreview.domain.DraftRecord>? = null
        var guides: List<com.ezevillo.gitreview.domain.GuideRecord>? = null
        var walkthrough: com.ezevillo.gitreview.domain.WalkthroughRecord? = null

        val hasFinishConflict = if (status.exitCode == 0) {
            try {
                parsePorcelain(status.stdout).finish != null
            } catch (_: Exception) {
                false
            }
        } else false

        if (status.exitCode == 2) {
            val list = invoker.invoke("list", listOf("--porcelain"), cwd)
            if (list.exitCode == 0 && !list.timedOut) {
                try {
                    // One invocation, two readings: the fixes branches travel in
                    // the same output as the inventory, so the footer section
                    // costs no extra process per refresh.
                    branches = parseListPorcelain(list.stdout)
                    fixes = parseListFixes(list.stdout)
                } catch (_: Exception) {
                    branches = emptyList()
                    fixes = emptyList()
                }
            }
            val cfg = invoker.invoke("config", listOf("--porcelain"), cwd)
            if (cfg.exitCode == 0 && !cfg.timedOut) {
                try {
                    val parsed = parseConfigPorcelain(cfg.stdout)
                    config = parsed.config
                    candidates = parsed.candidates
                    remotes = parsed.remotes
                    drafts = parsed.drafts
                    guides = parsed.guides
                    walkthrough = parsed.walkthrough
                } catch (_: Exception) {
                    // leave null
                }
            }
        }

        val hasFinishPending = branches.any { it.finish?.state == "pending" }
        val situation = situationFor(status.exitCode, hasFinishConflict, hasFinishPending)

        val state = when (situation) {
            Situation.REVIEW, Situation.FINISH_CONFLICT -> {
                try {
                    val porcelain = parsePorcelain(status.stdout)
                    ReviewState(
                        situation = situation,
                        state = porcelain.state,
                        entries = porcelain.entries,
                        files = porcelain.files,
                        branches = emptyList(),
                        subjects = porcelain.subjects,
                        authors = porcelain.authors,
                        base = porcelain.base,
                        finish = porcelain.finish,
                        readonly = porcelain.readonly,
                        keysOnly = porcelain.keysOnly,
                        draft = porcelain.draft,
                        draftPath = porcelain.draftPath,
                    )
                } catch (e: Exception) {
                    ReviewState(
                        situation = Situation.ERROR,
                        stderr = e.message ?: "failed to parse status porcelain",
                    )
                }
            }
            Situation.NO_REVIEW, Situation.FINISH_PENDING -> ReviewState(
                situation = situation,
                branches = branches,
                fixes = fixes,
                config = config,
                candidates = candidates,
                remotes = remotes,
                drafts = drafts,
                guides = guides,
                walkthrough = walkthrough,
            )
            Situation.OUT_OF_RANGE -> ReviewState(
                situation = situation,
                stderr = status.stderr,
            )
            else -> ReviewState(
                situation = situation,
                stderr = status.stderr.ifBlank { "status failed (exit ${status.exitCode})" },
            )
        }

        if (gen == generation.get()) current = state
        return state
    }

    companion object {
        /**
         * A version probe that ran out of time: the CLI is there and slow, not
         * absent, so what the panel says is where to look.
         */
        const val VERSION_TIMEOUT: String =
            "`git review --version` did not finish in time and was stopped. " +
                "Run it in a terminal to see how long it takes. " +
                "See the git review log in Help > Show Log."

        const val MULTI_ROOT_ERROR: String =
            "Open a single-folder workspace that is a git repository. git review uses " +
                "one root (like the CLI cwd); multi-root is not supported."
    }
}
